### ISSUE TITLE
Support garena taiwan skilltree

### ISSUE THREAD

**Issue by Chuanhsing** [EmmittJ/PoESkillTree#113]
1. PoeSkillTree.cs CharName change to read Skilltree.txt->main->out
2. PoeSkillTree.cs TreeAddress changable
location of chinese skilltree: http://web.poe.garena.tw/passive-skill-tree/ 
3. SkillTree.cs move BaseAttributes, _hybridAttributes, RenameImplicitAttributes to language file

**Comment by l0g0sys** [comment]
Can you please provide example of items JSON file from garena?
Guess, we should make some doc folder in repo to keep such things for now, and in future even some user manual (in HTML?). That way we would have not to clutter application's help with wall of text.

**Comment by Chuanhsing** [comment]
change the TreeAddress to http://web.poe.garena.tw/passive-skill-tree/ to download Skilltree.txt
but it will error when using CharName to search start node.

**Comment by EmmittJ** [comment]
I was just asked if we could so full localization for the Russian served as well. Does anyone know about localization and how difficult it it?

It would be really cool if we could do that. 

@l0g0sys @SpaceOgre

**Comment by l0g0sys** [comment]
UI localization shouldn't be much of problem. But full localization, i.e. ability to read data in non-EN languages and working with them might be quite of challenge :)
So what i'd like to see first, some example(s) of items JSON build files for both taiwanese and russian versions.

What @Chuanhsing mentioned is not enough to make it work with taiwanese data.

**Comment by EmmittJ** [comment]
Oh, I see. Since we use English as our main tree, we have hard coded
English strings like calculated stats and such. That could pose a pretty
big problem and a lot of rewriting.

**Comment by l0g0sys** [comment]
Yep, but that's solvable with bit of work.
Also, i'd like to know more about patch cycles. Are they behind EN version with patches or ahead? Are they really only localized versions of same data or they have also own items (e.g. unique gear, gems) ?

**Comment by EmmittJ** [comment]
Not sure about the patch cycle, but I think they are currently on the same
patch as US. As for items they should all be the same, but I don't play
there. I guess we will have to wait on that one.

**Comment by Ttxman** [comment]
Garena have english version on:
http://web.poe.garena.com/
@l0g0sys, @EmmittJ
According to patch notes they are on different patch rate and on different patch numbers
https://web.poe.garena.com/forum/view-forum/patch-notes
vs 
http://www.pathofexile.com/forum/view-forum/patch-notes
1.3.1g is 1.3.1e on garena and 18 days late :)

For UI localization I would recommend https://github.com/SeriousM/WPFLocalizationExtension. (https://wpflocalizeextension.codeplex.com/wikipage?title=Keys&referringTitle=Documentation). You cannot get localization more easy than:
```XML
<Button Content="{lex:Loc Test}" />
```
And one line labelled Test in resx file :)

**Comment by l0g0sys** [comment]
Ok, i've done some research on this and we start with UI localization first.

Found quite few libraries handling localization, but to be honest, they are all pretty much half assed and simply not suited for community based application. Either they handle translations only within XAML or only programatic translations within CS code. And all of them are more or less built around resrouce files being either embedded or compiled into satellite assemblies like native .NET localization does.

So, i've ended up making own stuff based on gettext tool to accomplish these features:
 1. Localization of both worlds (XAML and CS code).
 2. Automatized creation of template translations for easy "copy & translate" translation workflow.
 3. No project building required (i.e. you can do translation in your own installed application and then just submit localized catalog, no need to have development environment installed).

**Comment by EmmittJ** [comment]
Very cool man! I can't wait to see that :)

**Comment by Ttxman** [comment]
@l0g0sys you can implement your own localization provider for wpflocalizeextension and have all that functionality with few lines of code.

How are you handling the XAML part of localizations? Markup extension or some other way? BTW popups (and menus) are not part of parent window visual tree, it can break markup extensions that depends on visual parents.

**Comment by l0g0sys** [comment]
I see the fundamental idea of resource based translations in wpflocalizeextension flawed and pretty much same goes for the rest of C# libraries i found. But that's natural, because that's the way ppl were learned to do localization since early days of MFC C++.

If you deal with resource manager, you are limited to use property-like names as string identifiers. I don't want people to work with artificial identifiers like Localized_Resource_Property,  but rather to translate real stuff which will be actualy displayed.

Regarding XAML part, yep i use MarkupExtension, whole implementation is like 10 lines of code, so that's only functionality needed from wpflocalizeextension.

So if this is example of wpflocalizeextension:
```xml
<Button Content="{lex:Loc Key=Button_Label}" />
```
My implementation example looks like:
```xml
<Button>
  <l:Catalog Message="Close"/>
</Button>
```
Isn't it more natural?
As developer you don't have to think about:
  1. how to name the resource key
  2. or "didn't i already define some resource key for that label i should use, instead of creating new one?"

**Comment by Ttxman** [comment]
@logosys
Actually this will cause horrible things in XAML
You should add some markup extension like `{l:Catalog something}` your solution will add ~2 lines of XAML for each localized string. (transform self closing tag to pair tag and add line of content)
start with
```xml
<Button>
<l:Catalog Message="Close"/>
<Button>
```
vs
```xml
<Button Content={l:Catalog Close} />
```

Add tooltip
```xml
<Button Content={l:Catalog Close}  Tooltip={l:Catalog Close window} />
```
vs
```xml
<Button>
   <Button.Tooltip>
     <l:Catalog Message="Close window"/>
   </Button.Tooltip>
   <l:Catalog Message="Close"/>
<Button>
```

And add fairly common construction of localized string formats (countdown on button)
```xml
<Button Content="{Binding Source={StaticResource countdownTimer}}">
   <Button.Tooltip>
     <l:Catalog Message="Close window"/>
   </Button.Tooltip>
   <Button.ContentStringFormat>
     <l:Catalog Message="Countdown: {0}"/>
   </Button.ContentStringFormat>
<Button>
```

vs

```xml
<Button Content={l:Catalog Close}  Tooltip={l:Catalog Close window} 
ContentStringFormat="{l:Catalog Message=Countdown: {0}}" 
Content="{Binding Source={StaticResource countdownTimer}}"/>
```



Identifiers are used, because 1:1 or "string to string" translations just don't work in some cases. There are languages (probably most of them) where you cannot always use 1:1 translation of english, context can change the wording differently in different languages.

Good identifiers preserves context and looks like:
```
MainWindow_SaveConfirmation_caption
MainWindow_SaveConfirmation_text
```

Some examples why 1:1 translation is evil:
`Help` as in Mainmenu/help is `Nápověda` in Czech
`Help` as in 'help Karl' is "`Pomoci` Karlovi" in Czech
`Help` as in 'get help' is "sežeň `Pomoc`" in Czech

`OK` as in "Status:OK" is "Stav: `V pořádku`" in Czech
`OK` as in OK button is "`OK`" or "`Souhlasím`" (or maybe even "`Souhlas`" ) in Czech
`OK` as in OK/Cancel message dialog is "`Potvrdit`/Zrušit" in Czech

`Caption` as in "Headline", "Subtitle" or "Title" is "`Titulek`"  in Czech
`Caption` below image would be "`Popisek`" or "`Popis`"  in Czech
`Title` as in "Title after name: PhD.", is "`Titul`"  in Czech
`Title` as in "title in newspaper." (Heading), is "`Nadpis`"  in Czech


I am used to bad (or maybe forced) translations in application interface, caused by lack on context. Czech and English are in Indo-european language familly, and there were some interactions between them in history. It could be even worse in languages in different language families that differs more like Chinese (Sino-Tibetan language family) used by Taiwan garena.

**Comment by l0g0sys** [comment]
```xml
{l:Catalog something}
```
This markup cannot work. Due to use of real strings and not properites. That "something" must be property in this form not a string.
That's why {l:Catalog Message="string"} is used.

Extra lines really doesn't matter. If you are "line saver" just put it into 1 line... or even put whole document into one line if it bothers you :))) I've checked some XAML file and there are already object forms used instead of property form, so it's not a deal breaker.
And to be honest, those examples you wrote are more readable in object form than property form :)
Seriously... in these days to look at line counts before functionality and readability is foollish :)

No worries :) There will be translation contexts and plural forms.

Fixing typo actualy isn't a problem. There is fuzzy matching mechanism. You should really check the GNU gettext localization toolset.
And on the other hand, it gives higher chance of fixing typos as translators can point them out.
It's not like each day are changing walls of texts.

**Comment by Ttxman** [comment]
Actually you can use default value. Binding defaults unlabeled value as PATH, WPFlocalizeExtension defaults to Key. (I think you could do it with simple [ContentProperty("Message")] on your extension class.) I don't know how it handle spaces but it can be used :)

Extra lines are problem. You have to write them, auto formating is formating them for you (no document in one line). Annoingly long XAML is getting longer, XAML tree is getting deeper and editing the UI is getting more complicated.

Object forms are usually used when there is no other way, or if author is used to HTML, and they are often generated by Blend. You can write anything in XAML as object form (backgrounds, margins, fonts, alignment, ets), but most of them are usually used in property form.

I will not go to another argument about gettext, i haven't been convinced that it solves more problems that it introduces (in small projects) in life discussion so internet discussion is probably wasted time. It may be good tool to introduce translation to large completed project or some big project with tens of UI developers. But not for small scale project with few hundred translatable strings.

Plural forms aren't problem. The problem is incompatible (inconsistent) language features. For example: each noun and adjective in Czech language have 7 singular forms and 7 plural forms. Some of these forms are additionally modified based on gender of subject and used sentence structure. There isn't any remotely similar concept in English. And these inconsistencies will conflict differently between different languages and it will eventually break something

**Comment by l0g0sys** [comment]
Ok, let's get this discussion to some conclusion.
I'll finish my implementation, and if you guys won't like it, you can 
just discard it.

**Comment by EmmittJ** [comment]
Honestly, if it works and makes sense there is no reason to not use it. If
it is intuitive then there will be no problems with changing items or ui
design on the feature.

As for translations, we can only do the best we can. It will be almost
impossible to get everything right on the first iteration.

**Comment by Ttxman** [comment]
I will work fine. I just think it is unnecessary to implement our own localization system. 
You should at least make it a separate project, put it on your github and use it as dll. To give it some reusability.

we will know how much problems  there is and how big the problems are after someone try to create translation... :)

**Comment by l0g0sys** [comment]
Now help me how to translate Build to Czech or Slovak :)) Never played Czech/Slovak localized MMO.

**Comment by Ttxman** [comment]
There are some similarities in Russian (and other Slavic languages).
And we will probably need Russian and Taiwanese translations :)

**Comment by l0g0sys** [comment]
Lol, just checked Russian forum for kick ass and found... билд (which is 
"build" phonetically written in Russion alphabet :)
I'm going with "zostava" for now. I'll ask in my guild later, for now 
i'm not going to invent new word :)

**Comment by l0g0sys** [comment]
I've got to point to localize View Help which shows wall of text in pure XAML which would be crazy to translate like that.
So i think it would be much better to use some XAML control to display standalone document. Each language would have own localized document then.
Found two options:
 1. Use markdown notation document (like here on GitHub) then let it render with library like Markdown.Sharp from NuGet which provides XAML control to transform MD files into FlowDocument on the fly.
 2. Use WebView XAML control to display HTML. But, looking at description of it. It's damn robust with threading, marshaling and don't know what which could lead to some security issues.

I vote for (1).

**Comment by EmmittJ** [comment]
Honestly I would prefer option #1 as it is much easy to use.

**Comment by Ttxman** [comment]
option 1. WebView loads whole browser engine into memory it is overkill.

or just simple RichTextbox and RTF files :)

**Comment by l0g0sys** [comment]
Questio, Is there any reason why there is trailing space character in top-level menu items?
"_File ", "_Edit ", etc.

**Comment by Ttxman** [comment]
It's character for menu navigation ALT+(char_after '_'), its called accelerator key

**Comment by l0g0sys** [comment]
Lol, I was asking about extra SPACE character (ascii 32, 0x20). Not 
underscore character ;)

**Comment by Ttxman** [comment]
oh sorry i mixed up trailing and leading :) 
I think it is for added space between menu items. But it should be done in styling not in content, you can delete it.

**Comment by EmmittJ** [comment]
Pretty sure there isn't a reason :) @SpaceOgre would know for sure.

**Comment by l0g0sys** [comment]
Removed them and not much of difference.

Now another issue.
We have to remove PoEUrl It! buttom from bottom bar and move it to menu. Controls and labels there doesn't have enough space to allow sensible translation atm.
Where should it be moved? Tools or File menu?. It should not be in Edit as it doesn't change anything and neither in View as it doesn't control any part UI. It's really just a tool. But quite common place for export / import commands is File menu too.

**Comment by EmmittJ** [comment]
Hmm, I don't know if it handles external links links for RU and TW. I guess
tools would be the best place to put it if it has to be moved

**Comment by Ttxman** [comment]
You should change XAML for the the whole bottom panel. (and probably for most of the application)
Remove the positioning with margins. Put all the controls into grid with autosizing columns (Width="auto"). The URL should be in column with width="*" to span along remaining space.

Then controls will change size with translations, layout can stay same. Just let the UI resize with translated texts. And if you want buttons with same size just bind MinWidth to ActualWidth of the other one for both :)

**Comment by logosys** [comment]
If i knew how lol. UI design is not my cup of tea.
So i'll let someone else to do it.
Still it doesn't make sense to have export method at bottom. There will 
be more types of export in future and it's stupid to have 1 at bottom 
and other in menu. It has keyboard shortcut anyway for those who use it 
more than once per minute hehe.

On 8. 5. 2015 19:24, Ladislav Šeps wrote:
>
> You should change XAML for the the whole bottom panel. (and probably 
> for most of the application)
> Remove the positioning with margins. Put all the controls into grid 
> with autosizing columns (Width="auto"). The URL should be in column 
> with width="*" to span along remaining space.
>
> Then controls will change size with translations, layout can stay 
> same. Just let the UI resize with translated texts. And if you want 
> buttons with same size just bind MinWidth to ActualWidth of the other 
> one for both :)
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/EmmittJ/PoESkillTree/issues/113#issuecomment-100303801>.
>


-- 
Ďakujem,

Ing. Marek Grodný
grodny@oneclick.sk

**Comment by Ttxman** [comment]
like this: :)
https://github.com/Ttxman/PoESkillTree/commit/3be573e97c2b7d1eb944b7f181a0d786ff71328c

**Comment by SpaceOgre** [comment]
Regarding the space I think I just followed an tutorial when I added the menu and the space was there :) 
No clue as to why, this project is my first interaction with WPF.

**Comment by EmmittJ** [comment]
Currently don't have any plans to implement this nor have we for about a year so far. I don't think I can support two version of a skill tree especially on a language that I can't read.
